card/src/main/java/com/adyen/checkout/card/AddressDelegate.kt

```diff
@@ -0,0 +1,88 @@
+/*
+ * Copyright (c) 2022 Adyen N.V.
+ *
+ * This file is open source and available under the MIT license. See the LICENSE file for more info.
+ *
+ * Created by ozgur on 18/3/2022.
+ */
+
+package com.adyen.checkout.card
+
+import android.util.LruCache
+import com.adyen.checkout.card.api.AddressDataType
+import com.adyen.checkout.card.api.makeUrl
+import com.adyen.checkout.card.api.model.AddressItem
+import com.adyen.checkout.card.repository.AddressRepository
+import com.adyen.checkout.card.ui.AddressFormInput
+import com.adyen.checkout.components.base.Configuration
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.launch
+
+class AddressDelegate(
+    private val addressRepository: AddressRepository
+) {
+
+    private val _statesFlow: MutableStateFlow<List<AddressItem>> = MutableStateFlow(emptyList())
+    internal val statesFlow: Flow<List<AddressItem>> = _statesFlow
+
+    private val cache: LruCache<String, List<AddressItem>> = LruCache<String, List<AddressItem>>(CACHE_ENTRY_SIZE)
+
+    fun getStateList(
+        configuration: Configuration,
+        countryCode: String?,
+        coroutineScope: CoroutineScope
+    ) {
+        val addressSpecification = AddressFormInput.AddressSpecification.fromString(countryCode)
+        val needsStates = addressSpecification == AddressFormInput.AddressSpecification.BR ||
+            addressSpecification == AddressFormInput.AddressSpecification.US ||
+            addressSpecification == AddressFormInput.AddressSpecification.CA
+        if (!countryCode.isNullOrEmpty() && needsStates) {
+            val url = makeUrl(
```

Comment:
We are calling `makeUrl` 3 times now, 2 of them only to get the cache key. What if we use the state as cache key instead of the URL and only construct the url once? I know the intention was to use the same logic as the logo url but it just feels redundant to construct it twice 🤔